cpp/src/quantiles/quantiles.cu

```diff
@@ -0,0 +1,90 @@
+#include <memory>
+#include <vector>
+
+#include <cudf/types.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/detail/gather.cuh>
+#include <quantiles/quantiles_util.hpp>
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+template<typename SortMapIterator>
+std::unique_ptr<table>
+quantiles(table_view const& input,
+          SortMapIterator sortmap,
+          std::vector<double> const& q,
+          interpolation interp,
+          rmm::mr::device_memory_resource* mr)
+{
+    auto sortmap_lookup = [sortmap]
+        __device__(size_type idx) {
+            return *(sortmap + idx);
+        };
+
+    auto quantile_idx_lookup = [sortmap_lookup, interp, size=input.num_rows()]
+        __device__ (double q) {
+            return detail::select_quantile<size_type>(sortmap_lookup,
+                                                      size,
+                                                      q,
+                                                      interp);
+        };
+
+    rmm::device_vector<double> q_device = q;
+
+    auto quantile_idx_iter = thrust::make_transform_iterator(q_device.begin(),
+                                                             quantile_idx_lookup);
+
+    return detail::gather(input,
+                          quantile_idx_iter,
+                          quantile_idx_iter + q.size(),
+                          false,
+                          false,
+                          false,
+                          mr);
+}
+
+}
+
+std::unique_ptr<table>
+quantiles(table_view const& input,
+          std::vector<double> const& q,
+          interpolation interp,
+          cudf::sortedness sorted,
+          std::vector<order> const& column_order,
+          std::vector<null_order> const& null_precedence,
+          rmm::mr::device_memory_resource* mr)
+{
+    CUDF_EXPECTS(interp == interpolation::HIGHER ||
```

Comment:
Should `CUDF_EXPECT(q.size() > 0)` or `return empty_like(input)`?

Reply:
That sounds like an exception case to me.

Reply:
@harrism This is what I thought you were referring to when asking about empty _input_. With that in mind, I updated this case to return an empty table (for zero requested quantiles). I realize this has nothing to do with the other comment, but do you have an opinion here?